Rewrite Collapsible checkboxes as an ES Module

kr8n3r opened the pull request:
Make collapsible checkboxes component work better with screenreader users. Addresses https://trello.com/c/izqbbBLc/891-collapsible-components-work-poorly-for-screen-reader-users

## What

Primarily, changes needed to be made to the functionality as screenreader users were experiencing unexpected behaviour and labelling. Described in detail in the report, linked to the Trello ticket.

Secondly, we're in the process of migrating our legacy code to ES modules and as proposed changes in functionality are such that, while they could have be done in the original module, the module is a good candidate for a rewrite and removal
of dependency on JQuery along the way.

Functional changes

- Toggle button is now inserted directly before the checkboxes container and it instead of the fieldset is being toggled.

This was it follows the pattern accordion uses, where toggled content follows immediately in the DOM.
With this there is no longer the  need to auto focus on the group of checkboxes.

- Toggle button indicates collapsed / expanded state and text now includes legend text

Showing the state visually as well as making it clearer to screen reader users what this section is.

- Removal of "done" button and sticky footer

"Done" button was there to collapse the content and it was wrapped in a "sticky" container to be in view for mouse users to be able to close the section quicker on longer lists.

Removing the "done" button entirely addresses the concern of "...describing the done button as expanded indicates that I have selected that button and expanded the content...this is no the expected method of expanding content and it was very confusing."

The second issue with having a "done" or "collapse this section button" is that once triggered, there isn't a consistent next tabbabble element to place focus on, as this component is used in two different places where next form elements
differ. Placing the focus back on to the toggle button would not make sense either.

The removal of making the button sticky is mostly down to above removal of functionality.
While we could insert a none tabbable element there for visual users, the script itself is not written in a way that can be used is ES modules as well as in legacy code, where it still used.

There have been attempts to use Intersection Observer to make the toggle button visible in view on long lists, it would not be "sticky" at the bottom nor would it be aware and introduce scroll to avoid overlaying a focussed checkbox.

## Appearance

**Template folder management view**

<img width="424" alt="Screenshot 2024-08-29 at 12 15 49" src="https://github.com/user-attachments/assets/30b1e31a-cc68-44e1-bb99-edab825a489b">
<img width="437" alt="Screenshot 2024-08-29 at 12 15 54" src="https://github.com/user-attachments/assets/c77ff223-e519-4735-973a-e28278b8c560">
<img width="383" alt="Screenshot 2024-08-29 at 12 16 01" src="https://github.com/user-attachments/assets/8ff032f3-7477-4256-93df-47c4f2ba756e">


**User access management view**

<img width="388" alt="Screenshot 2024-08-29 at 12 14 42" src="https://github.com/user-attachments/assets/484bf06e-bd58-42eb-b09f-c528dc767104">
<img width="387" alt="Screenshot 2024-08-29 at 12 14 47" src="https://github.com/user-attachments/assets/b565dc71-0263-4e7a-b7f3-48d67a9a8e0b">


**No change to built static JS files with updated babel-core**
<img width="1060" alt="Screenshot 2024-08-19 at 12 11 34" src="https://github.com/user-attachments/assets/c1cd58d4-b2ee-484e-b5dd-403e9b027952">

kr8n3r committed:
New collapsible checkboxes ES module

This commit introduces an updated collapsible
checkboxes Javascript with functional and code
standard changes.

Why

Primarily, changes needed to be made to the
functionality as screenreader users were
experiencing unexpected behaviour and labelling.
Described in detail in the report, linked to this
ticket.

Secondly, we're in the process of migrating our
legacy code to ES modules and as proposed changes
in functionality are such that, while they could
have be done in the original module, the module is
a good candidate for a rewrite and removal
of dependency on JQuery along the way.

Functional changes

- Toggle button is now inserted directly before
the checkboxes container and it instead of the
fieldset is being toggled.

This was it follows the pattern accordion uses,
where toggled content follows immediately in the DOM.
With this there is no longer the  need to auto
focus on the group of checkboxes.

- Toggle button indicates collapsed / expanded state
and text now includes legend text

Showing the state visually as well as making it
clearer to screen reader users what this section is.

Removal of "done" button and sticky footer

"Done" button was there to collapse the content
and it was wrapped in a "sticky" container to be
in view for mouse users to be able to close the section
quicker on longer lists.

Removing the "done" button entirely addresses the
concern of "...describing the done button as
expanded indicates that I have selected
that button and expanded the content...this is no
the expected method of expanding content and it
was very confusing."

The second issue with having a "done" or "collapse
this section button" is that once triggered,
there isn't a consistent next tabbabble element
to place focus on, as this component is used
in two different places where next form elements
differ. Placing the focus back on to the toggle
button would not make sense either.

The removal of making the button sticky is mostly
down to above removal of functionality.
While we could insert a none tabbable element there
for visual users, the script itself is not written
in a way that can be used is ES modules as well
as in legacy code, where it still used.

There have been attepts to use Intersection
Observer to make the toggle button visible in view
on long lists, it would not be "sticky" at the
bottom nor would it be aware and introduce scroll
to avoid overlaying a focussed checkbox.

kr8n3r committed:
Update to collapsible checkboxes test

To be able to import and build ESM files, this
file's suffix has chnaged to .mjs and `require`
updated to import.

As for the updated tests, some have been removed
to reflect changed functionality. More on
functional changes in the new Javascript module
commit.

kr8n3r committed:
Lint .mjs files and fix

Set up Gulp to also check .mjs files for code style
standard. Some fixes needed to be made to adhere.

kr8n3r committed:
Toggle button indicator styles

Chevron arrow styles in the button
to indicate collapsed /expanded states.
Borrowed from
https://design-system.service.gov.uk/components/details/

kr8n3r committed:
Setup Jest to build & test ES modules files

As new/updated Javascript functionality is written
as ES modules, Jest needs to be able to transform
them to CommonJS as we're not in a position to be
fully ESM yet.

To transpile we use babel-jest which needs uptodate
babel-core and babel-present-env.

In package.json Jest config we tell it to transpile
.mjs test files as any new ESM modules code does not
need the window object legacy Javascript is attached
to, so any new tests are written with .mjs suffix.

To configure Babel, babel.config.js is set-up and
targets 'test' env as that's what Jest sets.

Updated babel-core and preset-env have not
introduced any extra code into the built static
`all.js` and `all-esm.mjs files` (see PR screenshot)

To be able to transpile ESM files,
`node --experimental-vm-modules node_modules/.bin/jest`
needs to be added to the the test command. More at
https://jestjs.io/docs/ecmascript-modules

kr8n3r commented:
@quis when you're back to speed after your holiday, if you have any thoughts on this, please feel free to comment

quis commented:
I’m ok with this, if we can use the same arrow style as the details component:
> [ ▶ Collapsed ]
>
> [ ▼ Expanded ]
> ☑️ 
> ☑️ 

But then I wonder if we shouldn’t just be using the details component anyway. Thoughts?

kr8n3r commented:
quick prototype using `details`
https://github.com/alphagov/notifications-admin/compare/collapsible-checkboxes-with-details?expand=1

![Screenshot 2024-08-28 at 12 10 37](https://github.com/user-attachments/assets/39e29fc1-d18f-4ba2-af86-d8cc1136371f)
![Screenshot 2024-08-28 at 12 10 47](https://github.com/user-attachments/assets/6455ceba-efa4-4738-badb-6901d2369f3a)
![Screenshot 2024-08-28 at 12 10 42](https://github.com/user-attachments/assets/57a360e5-f214-4677-ac2f-e2f411523261)

tombye commented on package.json:
```diff
@@ -31,8 +31,9 @@
   },
   "devDependencies": {
-    "@babel/core": "7.4.0",
+    "@babel/core": "^7.25.2",
```
Can we lock this down to an exact version number please, to match the others? Same with all others added in the same way.

tombye commented on tests/javascripts/collapsibleCheckboxes.test.mjs:
```diff
@@ -1,21 +1,14 @@
+import CollapsibleCheckboxes from '../../app/assets/javascripts/esm/collapsible-checkboxes.mjs'
+import * as helpers from './support/helpers.js'
 
 afterAll(() => {
```
Doesn't look like we need this `afterAll` now.

tombye reviewed (changes requested):
Spotted a few things, all commented, but this is some nice work overall and thanks for doing all the wrangling with jest and babel.

If we leave it as a button, not a `<details>`, I think we should make the arrow the same as the day/time picker:

<img width="371" alt="image" src="https://github.com/user-attachments/assets/49f29083-9772-4ea9-9f95-e24e975c0e7b">

## Things from cross browser testing

The button seems to start with spacing above and then move on the first click on Safari on MacOS and iOS. This is on MacOS:

![moving_button_macos](https://github.com/user-attachments/assets/0d108c34-f798-465f-9242-ed1cc40a0b86)

I noticed the folder icon for the summary still appearing on IE11. Not a game changer but didn't think this was intentional:

<img width="407" alt="image" src="https://github.com/user-attachments/assets/18438a35-570c-4803-acce-73eda9fc1f85">

## Suggestion for changing the order of the HTML

The other thing I'd suggest changing is to move the button out of the fieldset. It's currently a bit strange to hear the button after the legend:
- "[Group] Folders this team member can see"
- "Choose folders [button, expanded]"
- "[Unordered list]" (of checkboxes)

Moved outside, it's more like:
- "Choose folders [button, expanded]"
- "[Group] Folders this team member can see"
- "[Unordered list]" (of checkboxes)

...which I think makes more sense as the button label sets the expectation of folders and the group label following it meets that expectation.

kr8n3r committed:
Set selection summary icon visibility

As summary text and updating is done via JavaScript and only in supported browsers, we should not displaying it otherwise.

Makes use of the $govuk-frontend-supported-css-class that's set on the body element in supported browsers.

kr8n3r commented:
> Spotted a few things, all commented, but this is some nice work overall and thanks for doing all the wrangling with jest and babel.
> 
> If we leave it as a button, not a `<details>`, I think we should make the arrow the same as the day/time picker:
> 
> <img alt="image" width="371" src="https://private-user-images.githubusercontent.com/87140/362349856-49f29083-9772-4ea9-9f95-e24e975c0e7b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5Mjg3MzIsIm5iZiI6MTcyNDkyODQzMiwicGF0aCI6Ii84NzE0MC8zNjIzNDk4NTYtNDlmMjkwODMtOTc3Mi00ZWE5LTlmOTUtZTI0ZTk3NWMwZTdiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI5VDEwNDcxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiNGUwMjgyMGY1MWNiNDYzODc3NWY2NmVlMDE2YjAwZmI4ZDEyODFmZGZlNmU5YzM1M2FiYzNiN2I2ZmUxY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AYbLN1_OOLUnNuUXRl2IErV8vx-wNlH60iQQvH6LQ_c">
> ## Things from cross browser testing
> The button seems to start with spacing above and then move on the first click on Safari on MacOS and iOS. This is on MacOS:
> 
> ![moving_button_macos](https://private-user-images.githubusercontent.com/87140/362359575-0d108c34-f798-465f-9242-ed1cc40a0b86.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5Mjg3MzIsIm5iZiI6MTcyNDkyODQzMiwicGF0aCI6Ii84NzE0MC8zNjIzNTk1NzUtMGQxMDhjMzQtZjc5OC00NjVmLTkyNDItZWQxY2M0MGEwYjg2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI5VDEwNDcxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRkMmU4N2RiNTc1MzJmNjFmMTY4ZmU2ZTQyNTI1YjU3MzlmZGY2ZTlkNWVlMmQ5N2Q4OTE0NjZjNmYwMTBmMDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.I6YI532crv2geZ_h9w1zDZmanlfeOby8b7AGFspN5pQ) [ ![moving_button_macos](https://private-user-images.githubusercontent.com/87140/362359575-0d108c34-f798-465f-9242-ed1cc40a0b86.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5Mjg3MzIsIm5iZiI6MTcyNDkyODQzMiwicGF0aCI6Ii84NzE0MC8zNjIzNTk1NzUtMGQxMDhjMzQtZjc5OC00NjVmLTkyNDItZWQxY2M0MGEwYjg2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI5VDEwNDcxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRkMmU4N2RiNTc1MzJmNjFmMTY4ZmU2ZTQyNTI1YjU3MzlmZGY2ZTlkNWVlMmQ5N2Q4OTE0NjZjNmYwMTBmMDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.I6YI532crv2geZ_h9w1zDZmanlfeOby8b7AGFspN5pQ) ](https://private-user-images.githubusercontent.com/87140/362359575-0d108c34-f798-465f-9242-ed1cc40a0b86.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5Mjg3MzIsIm5iZiI6MTcyNDkyODQzMiwicGF0aCI6Ii84NzE0MC8zNjIzNTk1NzUtMGQxMDhjMzQtZjc5OC00NjVmLTkyNDItZWQxY2M0MGEwYjg2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI5VDEwNDcxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRkMmU4N2RiNTc1MzJmNjFmMTY4ZmU2ZTQyNTI1YjU3MzlmZGY2ZTlkNWVlMmQ5N2Q4OTE0NjZjNmYwMTBmMDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.I6YI532crv2geZ_h9w1zDZmanlfeOby8b7AGFspN5pQ) [ ](https://private-user-images.githubusercontent.com/87140/362359575-0d108c34-f798-465f-9242-ed1cc40a0b86.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5Mjg3MzIsIm5iZiI6MTcyNDkyODQzMiwicGF0aCI6Ii84NzE0MC8zNjIzNTk1NzUtMGQxMDhjMzQtZjc5OC00NjVmLTkyNDItZWQxY2M0MGEwYjg2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI5VDEwNDcxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRkMmU4N2RiNTc1MzJmNjFmMTY4ZmU2ZTQyNTI1YjU3MzlmZGY2ZTlkNWVlMmQ5N2Q4OTE0NjZjNmYwMTBmMDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.I6YI532crv2geZ_h9w1zDZmanlfeOby8b7AGFspN5pQ)
> 
> I noticed the folder icon for the summary still appearing on IE11. Not a game changer but didn't think this was intentional:
> 
> <img alt="image" width="407" src="https://private-user-images.githubusercontent.com/87140/362356765-18438a35-570c-4803-acce-73eda9fc1f85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5Mjg3MzIsIm5iZiI6MTcyNDkyODQzMiwicGF0aCI6Ii84NzE0MC8zNjIzNTY3NjUtMTg0MzhhMzUtNTcwYy00ODAzLWFjY2UtNzNlZGE5ZmMxZjg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI5VDEwNDcxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlMTZmOGUzODM5MDMyZjhjODI2YTBlYzM3YzFlZmJlZDI5NWRhY2M4MGQzMjcyMTFhOTFjODEwOWE2Yjk0MmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.U7x2fMInE7jSzv7b60K7npi6eNAXROjC7a1rvZ61688">
> ## Suggestion for changing the order of the HTML
> The other thing I'd suggest changing is to move the button out of the fieldset. It's currently a bit strange to hear the button after the legend:
> 
> * "[Group] Folders this team member can see"
> * "Choose folders [button, expanded]"
> * "[Unordered list]" (of checkboxes)
> 
> Moved outside, it's more like:
> 
> * "Choose folders [button, expanded]"
> * "[Group] Folders this team member can see"
> * "[Unordered list]" (of checkboxes)
> 
> ...which I think makes more sense as the button label sets the expectation of folders and the group label following it meets that expectation.

- updated arrow design
- not showing folder icon in non-supported browsers
- fixed Safari spacing
- moved the button before the fieldset

tombye reviewed (approved):
All issues fixed and it now makes much more sense when used with a screen reader due to the changes to the order of the elements. Nice work, ready to go now 👍🏻
